include/seqan3/core/algorithm/detail/execution_handler_parallel.hpp

```diff
@@ -189,23 +183,47 @@ class execution_handler_parallel
     {
         assert(state != nullptr);
 
-        if (!state->is_waiting)
+        state->stop_and_wait();
+    }
+
+private:
+    //!\brief An internal state stored on the heap to allow safe move construction/assignment of the class.
+    class internal_state
+    {
+    public:
+        /*!\name Constructors, destructor and assignment
+        * \brief Instances of this class are not copyable.
+        * \{
+        */
+        internal_state() = default; //!< Defaulted.
+        internal_state(internal_state const &) = delete; //!< Deleted.
+        internal_state(internal_state &&) = default; //!< Defaulted.
+        internal_state & operator=(internal_state const &) = delete; //!< Deleted.
+        internal_state & operator=(internal_state &&) = default; //!< Defaulted.
+
+        //!\brief Waits for threads to finish.
+        ~internal_state()
         {
-            state->is_waiting = true;
-            state->queue.close();
+            stop_and_wait();
+        }
+        //!\}
 
-            for (auto & t : state->thread_pool)
+        //!\brief Waits until all threads have been joined.
+        void stop_and_wait()
+        {
+            if (!is_waiting)
```

Comment:
I am wondering if the current mechanism is actually thread safe. 
So it could happen that one thread calls `stop_and_wait` and sets waiting to true. 
If now another thread is calling the destructor for some reason it will see that someone is already waiting but doesn't care and destructs the object. We can also not concurrently join the threads as this results in a data-race. And it is also possible that `is_waiting` is set to `true` after another thread has already entered the if block.
So we need to synchronise this i.e. only one thread can ever join the threads and all other wait until threads have been joined. I guess right now we only have a single producer so it is not really a problem but that might change in the future and then it will get tricky to figure out what is going on here.
So we could documented as not thread safe to call stop and wait and also document that the entire class is only intended for a single producer model. Then we could also throw away the is_waiting flag, because it suggests some safety that does not exixt